OnChain order cancellation

shoom3301 opened the pull request:
# Summary

## Changes
1. Button for selecting a cancellation type (on/off-chain). Only for wallets that support off-chain cancellation
2. Displaying of tx cost for on-chain cancellation (Tx in Goerli always UKNOWN, because we don't have a gas price here)
3. Added a link to a cancellation hash

<img width="980" alt="image" src="https://user-images.githubusercontent.com/7122625/231092868-4afb834b-073b-40e0-b74f-bd712cfd3a6f.png">

| | |
|--------|--------|
| <img width="507" alt="image" src="https://user-images.githubusercontent.com/7122625/230857719-7d83d05d-99c7-4c21-b93b-e4647354a421.png"> | <img width="497" alt="image" src="https://user-images.githubusercontent.com/7122625/230857888-2503b271-371a-4b14-b8eb-aedcf97cec3d.png"> |

> For code-reviewers: It will be easier to review the code by commits

  # To Test
  
### Smart contract wallet

1. Open CowSwap with a smart contract wallet (Gnosis Safe or other)
2. Create a limit order with a price above market
3. Click "Cancel order" in the context menu in the orders table
- [ ] there should NOT be an option to change the cancellation type
4. Click "Request cancellation"
- [ ] an order must be canceled using on-chain transaction

### Other wallets

1. Open CowSwap with a NOT smart contract wallet (Metamask or other)
2. Create a limit order with a price above market
3. Click "Cancel order" in the context menu in the orders table
- [ ] There should be an option to change the cancellation type (on-chain/off-chain)
4. Click "Request cancellation"
- [ ] an order must be canceled in accordance with selected type

### Eth-flow orders
- [ ] Canecalltion of eth-flow orders should work as before

anxolin reviewed (commented):
I manage to cancel it, but still the UX needs to be a bit polished. 


Once canceled, it shouldn't show you the cancel button. 
![image](https://user-images.githubusercontent.com/2352112/230919405-6fecbe22-4abb-4a25-9fc8-f3e07b5a8007.png)


~It didn't track the "canceling..." state. I would think that we should be "canceling..." when we already posted the TX. Also, maybe nice to have a way to see the cancelation tx?~
EDIT: on my second attempt, it did. Maybe I just didn't test correctly.

~One trivial implementation of the above is to add the tx to the list of tx we show. Although not sure that's the best UX. We could discuss this~

Not needed if we just handle the "Canceling..." state right. Although the part of showing the cancelation TX somewhere might be nice addition if its not hard.

~Also, when the state changed from "Open" to canceled, the UI showed "Canceling..." instead of "Canceled" (now should be final, not like in offchain cancelation that its possible to execute later)~

On my second attempt, it didn't happen. Maybe I made it up??

shoom3301 commented:
@elena-zh 
1. Off-chain cancellation is not a 100% guarantee. Potentially, an order might be filled even after off-chain cancellation. On-chain cancellation is 100% guaranteed.
2. Oops, will fix it
3. Yes, it's possible. Let's discuss it with the team. Anyway, we can improve it after merging this PR (I guess it should be fixed on the backend side)
4. Hmmm, I'm nott aware about this case, let's discuss it
5. Will check it
6. o_0 it shouldn't be possible
7. Out of scope

shoom3301 commented:
@elena-zh 
1. We discussed
2. Fixed
3. I couldn't reproduce. Anyway, I don't think we can fix it on frontend side
4. Fixed. Now you can't cancel an order while it in Signing state
5. I don't know how to implement it. Let's create another issue for it
6. Fixed
7. Out of scope

elena-zh commented:
Hey @shoom3301 , 

1. How to reproduce case 3: see the [video](https://www.loom.com/share/18c2581b29324b06a3f8f6238e447b8f)
2. Still I can see 'Cancel' order button for Cancelled (offchain), Expired, Cancelling orders when connected to EOA
![cancel for cancelled](https://user-images.githubusercontent.com/70885163/232054451-61b45551-7552-467b-b002-38444b622ba3.png)
3. I have noticed that when I sign offchain cancel order tx, the order goes to cancelling state, then can be returned into the Open state, then again can go to 'Cancelling' state. You can see it on the video in p1. 
4. I think, there is no need to show Cancel TX link for the ETH orders, as it presents on the progress bar. WDYT?
![cancelling](https://user-images.githubusercontent.com/70885163/232055093-de09fbaa-d65a-41d9-9d79-534f32a36ccf.png)
4. Cancel order link (when connected to SC wallets) still navigates to the 404 page
![not correct](https://user-images.githubusercontent.com/70885163/232055307-3ed6eaa2-9c68-485a-b908-79efb3603a60.png)
https://goerli.etherscan.io/tx/0x3678d1b1b3ea016158f3aa053cf9d1a3f7ed26644823fb30d2ef229d2f58c692
Just a bit more context: maybe we could copy the link logic that is implemented for this component?
![image](https://user-images.githubusercontent.com/70885163/232055941-cc082430-9c24-4f47-8762-abe7e097e7ea.png)


6. Cancellation TX link is still displayed for Filled orders (SC wallet)
![6 not removed](https://user-images.githubusercontent.com/70885163/232055602-729c5892-b8b3-40b2-8572-6894e1f40b62.png)

Thanks

anxolin commented on src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts:
```diff
@@ -0,0 +1,120 @@
+import { COW } from 'constants/tokens'
+
+const chainId = 1
+const settlementCancellationTxHash = '0xcfwj23g4fwe111'
```
Lol, it looks like you hit your keyboard hard to randomize this address

shoom3301 replied:
Ha ha :) Yes, why not?

anxolin commented on src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts:
```diff
@@ -0,0 +1,120 @@
+      expect(setOrderCancellationHash.mock.calls[0]).toMatchSnapshot()
+    })
+  })
+})
```
Vert cool tests 🔥

anxolin commented on src/cow-react/common/utils/isOrderCancellable.test.ts:
```diff
@@ -0,0 +1,58 @@
+    const order = {
+      inputToken: COW[1],
+      status: OrderStatus.PENDING,
+      isCancelling: true, // <-----
```
hehe, nice comment. Does the trick :)

anxolin commented on src/cow-react/common/utils/isOrderCancellable.test.ts:
```diff
@@ -0,0 +1,58 @@
+    } as Order
+
+    expect(isOrderCancellable(order)).toBe(false)
+    expect(isOrderCancellable({ ...order, isCancelling: false })).toBe(true)
```
nit: not sure why you test the negative too inside this test. In principle, that should be other test

shoom3301 replied:
Corresponding to the best practices - yes, it should. But the current version is two times shorter and the readability is not worse, isn't?

anxolin commented on src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:
```diff
@@ -8,12 +9,14 @@ import OrderCancelledImage from 'assets/cow-swap/order-cancelled.svg'
 import { SafeMultisigTransactionResponse } from '@gnosis.pm/safe-service-client'
 import { getActivityState } from 'hooks/useActivityDerivedState'
 import { CancelButton } from '@cow/common/pure/CancelButton'
+import { ExplorerDataType, getExplorerLink } from '@src/utils/getExplorerLink'
```
do you need the `@src`?

shoom3301 replied:
Nope, deleted it, thanks!

anxolin reviewed (approved):
This is a great great PR :)

shoom3301 commented:
@elena-zh thanks!
1. Could you open an issue for the case please?
2. Fixed
3. See p1
4. I hide it when an order in "open" state
5. Fixed, thanks for the reference!
6. Fixed

alfetopito commented on src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:
```diff
@@ -8,17 +9,18 @@ import OrderCancelledImage from 'assets/cow-swap/order-cancelled.svg'
 export function GnosisSafeLink(props: {
   chainId: number
   safeTransaction?: SafeMultisigTransactionResponse
-  gnosisSafeThreshold: number
```
Was this not used at all?

shoom3301 replied:
Yes

alfetopito reviewed (approved):
Very nice!

A few questions on the code

shoom3301 commented on src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.ts:
```diff
@@ -0,0 +1,33 @@
+
+      if (receipt?.hash) {
+        cancelPendingOrder({ id: order.id, chainId })
+        addTransaction({ hash: receipt.hash, ethFlow: { orderId: order.id, subType: 'cancellation' } })
```
Nice catch! Thank you! Added a condition

elena-zh commented:
Hey @shoom3301 , 
for the cases 1 and 3 I opened #2235 issue
Cases 2, 4 LGTM!
Cases 5 and 6 are still reproducible to me. 
Could you please take a look at them once again? 

Thanks

shoom3301 commented:
@elena-zh 5,6 cases work well to me.
I guess it's a cache problem. Please, clear your localStorage and try again with new orders.

alfetopito commented on src/custom/state/orders/updaters/GpOrdersUpdater.ts:
```diff
@@ -103,9 +103,10 @@ function _transformGpOrderToStoreOrder(
+  // It can happen that EthFlow cancellation is identified in the app before the API is aware
   // In that case
-  if (order.status === 'cancelled') {
+  if (order.ethflowData && order.status === 'cancelled') {
```
Didn't work?

shoom3301 replied:
It changed the behavior for off-chain cancelled orders :(
